# Labor Rights RAG Assistant

### Project Goal:

This project implements a Retrieval-Augmented Generation (RAG) pipeline for answering legal questions based on the Guide to Employment Rights in Lebanon.

## Imports

In [2]:
import os
import pickle
import numpy as np
import pandas as pd
from tqdm import tqdm
from dotenv import load_dotenv
from openai import OpenAI

import faiss
from sentence_transformers import SentenceTransformer

from langchain_community.document_loaders import PyPDFDirectoryLoader
from langchain_text_splitters import RecursiveCharacterTextSplitter

## Setup 

In [3]:
load_dotenv()

# Set API key 
OPENAI_API_KEY = os.getenv("OPENAI_API_KEY")
if not OPENAI_API_KEY:
    raise ValueError("OPENAI_API_KEY not found in .env file")

# Define file path
DATA_DIR = "data"

if not os.path.exists(DATA_DIR):
    raise FileNotFoundError(f"{DATA_DIR} directory not found")

print("Environment loaded successfully.")

Environment loaded successfully.


### Pipeline Overview:

1. Document Loading & Chunking:

    - The PDF is loaded and split into overlapping text chunks.

    - Chunk size and overlap are chosen to balance context coverage and retrieval precision.

2. Embedding Generation

    - Each chunk is converted into a dense vector using a sentence-transformer model.

    - Embeddings are normalized to enable cosine similarity via inner product.

3. Vector Storage (FAISS)

    - Embeddings are stored in a FAISS IndexFlatIP index.

    - This enables efficient similarity search.

4. Query Processing

    - The user query is embedded.

    - Top-k most similar chunks are retrieved.

5. Answer Generation

    - Retrieved chunks are injected into a structured prompt.

    - The LLM (gpt-4.1-mini) generates a grounded answer.

    - If context is insufficient, the system explicitly refuses to hallucinate.

## PDF Loading

In [4]:
loader = PyPDFDirectoryLoader(DATA_DIR)
documents = loader.load()

if not documents:
    raise ValueError("No documents were loaded. Check your data folder.")

print(f"Total pages loaded (including blanks): {len(documents)}")

# Remove empty pages
documents = [doc for doc in documents if doc.page_content.strip()]

print(f"Pages after removing blanks: {len(documents)}")

if not documents:
    raise ValueError("All pages are empty after cleaning.")

# Show sample from first valid page
sample = documents[44]

print("\nSample Preview:")
print(sample.page_content[:500])

print("\nMetadata:", sample.metadata)

Total pages loaded (including blanks): 96
Pages after removing blanks: 95

Sample Preview:
CREDITS 
Acknowledgements:
Legal guide authored by Layal Abou Daher, legal consultant. Edited by Martin Clutterbuck/NRC. January 2024.
Design:
 
All photos @NRC and other external photographers as credited. 
Cover Photo: Palestinian Baker, (Photo: Racha El Daoi/NRC)
Back Cover Photo: Making Falafel, Bekaa (Photo: Racha El Dao/NRC)

Metadata: {'producer': 'Adobe PDF Library 17.0', 'creator': 'Adobe InDesign 19.0 (Windows)', 'creationdate': '2024-01-09T12:50:11+03:00', 'moddate': '2024-01-09T13:37:41+03:00', 'trapped': '/False', 'source': 'data\\Self-Employment-Guide.pdf', 'total_pages': 52, 'page': 1, 'page_label': '2'}


In [15]:
set(doc.metadata["source"] for doc in documents)

{'data\\Employment-Rights-Guide.pdf', 'data\\Self-Employment-Guide.pdf'}

## Chunking

In [5]:
import tiktoken

# Token counter
encoding = tiktoken.get_encoding("cl100k_base")

def token_length(text):
    return len(encoding.encode(text))

splitter = RecursiveCharacterTextSplitter(
    chunk_size=400,       # tokens 
    chunk_overlap=80,
    length_function=token_length,
    separators=["\n\n", "\n", ".", " ", ""]
)

chunks = splitter.split_documents(documents)

# Add chunk ID
for i, chunk in enumerate(chunks):
    chunk.metadata["chunk_id"] = i
    
print(f"Total chunks created: {len(chunks)}")
print(f"Created {len(chunks)} chunks from {len(documents)} documents")

print(f"\nSample Chunk:")
print(chunks[0].page_content[:300])
print(f"\nChunk metadata: {chunks[0].metadata}")

Total chunks created: 202
Created 202 chunks from 95 documents

Sample Chunk:
Acknowledgements. Legal guide authored by NRC and El Meouchi law firm ( https://www.elmeouchi.
com/  Chadia El Meouchi and Elie El Feghali), facilitated by TrustLaw, the Thomson Reuters Foundation’s 
global pro bono service. Edited by Martin Clutterbuck/NRC. Special thanks to Lianna Badamo, Sarah 
G

Chunk metadata: {'producer': 'Adobe PDF Library 17.0', 'creator': 'Adobe InDesign 18.1 (Windows)', 'creationdate': '2023-03-16T15:35:55+03:00', 'moddate': '2023-03-18T09:26:14+03:00', 'trapped': '/False', 'source': 'data\\Employment-Rights-Guide.pdf', 'total_pages': 44, 'page': 1, 'page_label': '2', 'chunk_id': 0}


## Generate Embeddings

In [6]:
# Initialize embedding model
embedder = SentenceTransformer('all-MiniLM-L6-v2')

# Extract text from chunks
texts = [chunk.page_content for chunk in chunks]

# Generate embeddings
embeddings = embedder.encode(
    texts,
    show_progress_bar=True,
    batch_size=32,
    normalize_embeddings=True
)

dimension = embeddings.shape[1]

print(f"Embedding dimension: {dimension}")
print(f"Total embeddings: {embeddings.shape[0]}")

Loading weights: 100%|██████████| 103/103 [00:00<00:00, 251.76it/s, Materializing param=pooler.dense.weight]                             
BertModel LOAD REPORT from: sentence-transformers/all-MiniLM-L6-v2
Key                     | Status     |  | 
------------------------+------------+--+-
embeddings.position_ids | UNEXPECTED |  | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.
Batches: 100%|██████████| 7/7 [00:14<00:00,  2.09s/it]

Embedding dimension: 384
Total embeddings: 202


## Create FAISS Index

In [7]:
index = faiss.IndexFlatIP(dimension)  # use cosine similarity
index.add(embeddings.astype("float32"))

print(f"FAISS index created with {index.ntotal} vectors.")

FAISS index created with 202 vectors.


## Save FAISS Index

In [8]:
import pickle

faiss.write_index(index, "index.faiss")

with open("chunks.pkl", "wb") as f:
    pickle.dump(chunks, f)

print("Vector store saved successfully.")

Vector store saved successfully.


## Retrieval and Answer Generation Functions

In [43]:
from openai import OpenAI
from dotenv import load_dotenv
import os

load_dotenv()
client = OpenAI(api_key=os.getenv("OPENAI_API_KEY"))

def retrieve(query, k=3):
    query_embedding = embedder.encode(
        [query],
        normalize_embeddings=True
    )

    scores, indices = index.search(
        query_embedding.astype("float32"),
        k
    )

    results = []
    for idx, score in zip(indices[0], scores[0]):
        metadata = chunks[idx].metadata
        results.append({
            "score": float(score),
            "source": metadata.get("source"),
            "page": metadata.get("page_label"),
            "content": chunks[idx].page_content
        })

    return results


def generate_answer(query, k=3):
    try:
        retrieved_docs = retrieve(query, k=k)

        if not retrieved_docs:
            return {
                "answer": "No relevant documents were retrieved.",
                "sources": []
            }

        context = "\n\n---\n\n".join(
            doc["content"] for doc in retrieved_docs
        )

        if not context.strip():
            return {
                "answer": "Retrieved documents contain no usable content.",
                "sources": []
            }

        prompt = f"""
You are a legal assistant answering questions strictly based on the provided context.

Rules:
- Only use the information in the context.
- If the answer is not in the context, say: "The guide does not provide this information."
- Be clear and structured.
- Do not invent laws or numbers.

Context:
{context}

Question:
{query}

Answer:
"""

        response = client.chat.completions.create(
            model="gpt-4.1-mini",
            messages=[
                {"role": "system", "content": "You are a precise legal assistant."},
                {"role": "user", "content": prompt}
            ],
            temperature=0,
            max_tokens=500
        )

        if not response.choices:
            return {
                "answer": "Model returned no response.",
                "sources": []
            }

        answer = response.choices[0].message.content.strip()

        if "The guide does not provide this information." in answer:
             return {
        "answer": answer,
        "sources": []
            }

        sources = list(set(
            f"{doc['source']} (Page {doc['page']})"
            for doc in retrieved_docs
        ))

        return {
            "answer": answer,
            "sources": sources
        }

    except Exception as e:
        return {
            "answer": f"An error occurred during answer generation: {str(e)}",
            "sources": []
        }

## Test RAG Pipeline on Sample Questions

In [57]:
evaluation_questions = [

    # EMPLOYMENT RIGHTS GUIDE (17)
    "What is the legally mandated number of annual leave days?",
    "How does annual leave increase based on years of service?",
    "What is the official minimum wage?",
    "What are the maximum legal working hours per week?",
    "How is overtime compensated?",
    "What are the rules governing sick leave?",
    "Is maternity leave paid and for how long?",
    "What rights exist during the probation period?",
    "What prior notice is required before termination?",
    "What compensation is due in case of abusive dismissal?",
    "Under what conditions can employment be terminated due to force majeure?",
    "What are the employer's obligations regarding social security contributions?",
    "Are foreign workers entitled to social security benefits?",
    "Are employees protected from dismissal while on leave?",
    "What are the procedures for challenging abusive dismissal?",
    "Does the guide regulate cryptocurrency salary payments?",
    "Does the guide contain rules about remote work or work-from-home policies?",
    
    # SELF-EMPLOYMENT & SMALL BUSINESS GUIDE (12)
    "How does the guide define workers not linked to an employer?",
    "Are self-employed individuals covered by social security?",
    "Are foreigners eligible for social security benefits?",
    "When is a daily worker considered an employee versus self-employed?",
    "Must individuals conducting commercial activities register as traders?",
    "What activities are considered commercial in nature?",
    "Are workers who work for multiple employers covered by social security?",
    "Are self-employed individuals required to pay tax?",
    "Are self-employed individuals required to pay municipal fees, and on what basis are they calculated?",
    "Are small-scale daily labourers considered traders under commercial law?",
    "Does the guide set minimum pricing standards for freelance services?",
    "Does the guide provide government grants or startup funding schemes for self-employed workers?"
]

In [58]:
import pandas as pd

results = []

for question in evaluation_questions:
    output = generate_answer(question, k=3)

    results.append({
        "question": question,
        "answer": output["answer"],
        "sources": ", ".join(output["sources"])
    })

In [59]:
# For clean display

from IPython.display import display, Markdown

for r in results:
    display(Markdown(f"### Question\n{r['question']}"))
    display(Markdown(f"**Answer:**\n\n{r['answer']}"))
    display(Markdown(f"**Sources:** {r['sources']}"))
    display(Markdown("---"))

### Question
What is the legally mandated number of annual leave days?

**Answer:**

The legally mandated number of annual leave days according to Article 39 of the Labour Code is 15 working days, fully paid, provided that the employee has been employed by the employer for at least one year. 

However, while the Labour Code itself does not require an increase in annual leave based on years of service, the Ministry of Labour suggests the following minimum number of annual leave days based on length of service:

- 15 working days for employees employed between 1 to 5 years.
- 17 working days for employees employed between 5 to 10 years.
- 19 working days for employees employed between 10 to 15 years.
- 21 working days for employees employed for more than 15 years.

These suggested increases are based on Lebanon's membership in several Arab and International conventions.

**Sources:** data\Employment-Rights-Guide.pdf (Page 24), data\Employment-Rights-Guide.pdf (Page 16)

---

### Question
How does annual leave increase based on years of service?

**Answer:**

Annual leave increases based on years of service as follows:

- For employees working between 1 to 5 years: 15 working days of annual leave.
- For employees working between 5 to 10 years: 17 working days of annual leave.
- For employees working between 10 to 15 years: 19 working days of annual leave.
- For employees working more than 15 years: 21 working days of annual leave.

This increase is suggested by the Ministry of Labour when registering a new internal regulation, although the Labour Code itself only mandates a minimum of 15 days after one year of employment.

**Sources:** data\Employment-Rights-Guide.pdf (Page 24), data\Employment-Rights-Guide.pdf (Page 16)

---

### Question
What is the official minimum wage?

**Answer:**

The official minimum wage in Lebanon is currently LBP 675,000 per month for private employers. Additionally, a ‘cost of living allowance’ has been added by government decree, bringing the total minimum wage up to LBP 2,000,000 per month.

**Sources:** data\Employment-Rights-Guide.pdf (Page 13), data\Employment-Rights-Guide.pdf (Page 31)

---

### Question
What are the maximum legal working hours per week?

**Answer:**

The maximum legal working hours per week are 48 hours, although there are exceptions for overtime.

**Sources:** data\Employment-Rights-Guide.pdf (Page 11), data\Employment-Rights-Guide.pdf (Page 15), data\Employment-Rights-Guide.pdf (Page 12)

---

### Question
How is overtime compensated?

**Answer:**

Overtime compensation is governed by the following rules based on the provided context:

1. Employers must inform employees within 24 hours of the situation requiring overtime and the time needed to complete the work.

2. The salary for additional working hours (overtime) is increased by 50% compared to the regular working hours.

3. Working hours may be increased for certain businesses such as restaurants or cafes.

4. There are no specific provisions governing overtime within the Labour Code itself.

Therefore, while the Labour Code lacks specific overtime provisions, the salary for overtime work is increased by 50% compared to regular hours, and employees must be informed promptly about the overtime work.

**Sources:** data\Employment-Rights-Guide.pdf (Page 22), data\Employment-Rights-Guide.pdf (Page 15), data\Employment-Rights-Guide.pdf (Page 23)

---

### Question
What are the rules governing sick leave?

**Answer:**

Rules Governing Sick Leave in Lebanon:

1. Entitlement:
   - Employees who have worked between 3 months and 2 years are entitled to sick leave as follows:
     - Up to half a month of full pay.
     - Up to half a month of half pay.
   - Sick leave entitlements increase progressively for employees who have worked longer periods: up to 4 years, up to 6 years, up to 10 years, and more than 10 years.

2. Medical Report:
   - Sick leave must be supported by a medical report.
   - The employer has the right to verify the medical report.

3. Maximum Limit:
   - Sick leave can be claimed until the maximum limit is reached.

4. Impact on Annual Leave:
   - Employers may reduce the annual leave by 8 days if an employee’s sick leave exceeds one month within a year.

5. Protection from Dismissal:
   - Employees cannot be dismissed or served with a dismissal notice while on sick leave.

References:
- Article 40 and Article 41 of the Labour Code.
- Guide to Employment Rights, Section on Sick Leave.

**Sources:** data\Employment-Rights-Guide.pdf (Page 24), data\Employment-Rights-Guide.pdf (Page 3), data\Employment-Rights-Guide.pdf (Page 17)

---

### Question
Is maternity leave paid and for how long?

**Answer:**

Maternity leave in Lebanon is paid and lasts for a period of seven (7) weeks, which includes the period before and after giving birth. Wages and salary entitlements shall be paid in total to women during their maternity leave.

**Sources:** data\Employment-Rights-Guide.pdf (Page 14), data\Employment-Rights-Guide.pdf (Page 17), data\Employment-Rights-Guide.pdf (Page 26)

---

### Question
What rights exist during the probation period?

**Answer:**

During the probation period, the following rights exist:

1. **Inclusion in Contracts**: Probation periods can be included in both fixed-term and open-ended employment contracts.

2. **Duration**: The probation period is set for three months under the Labour Code. It is of public order and cannot be extended.

3. **Termination Rights**: Both the employer and the employee have the right to terminate the employment contract during the probation period without any prior notice or indemnity.

4. **Ease of Termination**: It is easier to terminate an employment contract during the probation period than after it has expired.

These rights are established under Articles 50(d) and 74 of the Labour Code.

**Sources:** data\Employment-Rights-Guide.pdf (Page 3), data\Employment-Rights-Guide.pdf (Page 12), data\Employment-Rights-Guide.pdf (Page 20)

---

### Question
What prior notice is required before termination?

**Answer:**

The prior notice required before termination depends on the duration of the employment contract as follows:

- One (1) month prior notice if the contract was for three (3) years or less.
- Two (2) months prior notice if the contract was for more than three (3) years but less than six (6) years.
- Three (3) months prior notice if the contract was for more than six (6) years but less than twelve (12) years.
- Four (4) months prior notice if the contract was for twelve (12) years or more.

Additional requirements:
- The notice must be delivered in writing and personally to the concerned party.
- The party receiving the notice has the right to ask for clarification on the causes of the termination if these were not specified in the written notice.

**Sources:** data\Employment-Rights-Guide.pdf (Page 19), data\Employment-Rights-Guide.pdf (Page 21)

---

### Question
What compensation is due in case of abusive dismissal?

**Answer:**

Compensation in Case of Abusive Dismissal:

1. **Right to Challenge and Reinstatement**  
   - Workers terminated without valid reason (abusive termination) may challenge the dismissal before the Labour Arbitration Council within 30 days.  
   - If the dismissal is found unjustified, the employer must reinstate the employee.  

2. **Indemnity Calculation**  
   - The indemnity depends on:  
     - Type of work  
     - Employee’s age  
     - Years of service  
     - Medical and family status  
     - Size of the damage and intensity of rights abuse in the case of abusive termination  
   - The indemnity ranges between **two (2) and twelve (12) months’ salary**, in addition to the legal indemnity due for termination of service.  

3. **Additional Indemnity for Failure to Comply with Notice Requirements**  
   - If the employer fails to provide written notice specifying the causes of termination or fails to deliver the notice personally, they must pay an indemnity equivalent to the salary for the notice period.  

Summary: In case of abusive dismissal, the employee is entitled to reinstatement and an indemnity ranging from 2 to 12 months’ salary plus the legal indemnity for termination, with possible additional indemnity for failure to comply with notice provisions.

**Sources:** data\Employment-Rights-Guide.pdf (Page 24), data\Employment-Rights-Guide.pdf (Page 21), data\Employment-Rights-Guide.pdf (Page 20)

---

### Question
Under what conditions can employment be terminated due to force majeure?

**Answer:**

Under the Labour Code in Lebanon, employment can be terminated due to force majeure under the following conditions:

1. The employer may terminate some or all ongoing employment contracts in the institution in case of:
   - Force majeure,
   - Economic or technical circumstances, such as:
     - Reduction of the size of the institution,
     - Substitution of a certain production system by another,
     - Complete cessation of work of the institution.

2. The employer must:
   - Inform the Ministry of Labour of the intention to terminate the contracts one month prior to the effective termination,
   - Discuss with the Ministry to establish a final program for termination, considering:
     - Seniority of employees,
     - Specialization,
     - Age,
     - Familial and social status,
     - Necessary means for re-employment.

3. Workers laid off due to force majeure have priority for reclaiming their jobs if the company resumes operations within a year.

4. In the context of COVID-19, the pandemic may be considered an event of force majeure justifying suspension (not termination) of contracts. However, termination on this basis requires compliance with Article 50(F) of the Labour Code and Ministry notification.

5. Courts may be lenient but may also consider termination due to temporary impossibility to perform work as abusive dismissal, requiring compensation.

Therefore, termination due to force majeure must follow the legal procedure of notifying and consulting the Ministry of Labour and take into account employee protections to avoid abusive dismissal claims.

**Sources:** data\Employment-Rights-Guide.pdf (Page 22), data\Employment-Rights-Guide.pdf (Page 23), data\Employment-Rights-Guide.pdf (Page 20)

---

### Question
What are the employer's obligations regarding social security contributions?

**Answer:**

The employer's obligations regarding social security contributions are as follows:

1. **Registration**: Employers are required to register all employees at the National Social Security Fund (NSSF).

2. **Monthly Contributions**: Employers must pay monthly contributions to the NSSF for each employee, calculated based on the employee’s salary:

   - **Family Coverage**: 6% of the employee’s salary, fully borne by the employer. This applies to all employees, including foreigners who do not benefit from the family allowance.
   
   - **Sickness Coverage**: 8% of the employee’s salary (part of the total 11% sickness coverage contribution), with the employee paying the remaining 3%. This applies to all employees, including foreigners who do not benefit from the sickness coverage.
   
   - **End-of-Service Indemnity**: 8.5% of the employee’s salary, except for Syrian employees who are exempt as they do not benefit from this indemnity.

3. **Additional Calculation for High Cost of Living**: Since 1 April 2022, for the purpose of calculating NSSF contributions, LBP 1,325,000 must be added to the minimum monthly wage (making it LBP 2,000,000) and to the basic monthly wage of employees earning LBP 4,000,000 or less as of 1 March 2022.

In summary, the employer must register employees with the NSSF and pay the specified percentages of the employee’s salary for family coverage, part of sickness coverage, and end-of-service indemnity (except for Syrians), using the adjusted wage base where applicable.

**Sources:** data\Employment-Rights-Guide.pdf (Page 14), data\Employment-Rights-Guide.pdf (Page 11), data\Self-Employment-Guide.pdf (Page 27)

---

### Question
Are foreign workers entitled to social security benefits?

**Answer:**

Based on the provided context:

Foreign workers in Lebanon are required to contribute to the National Social Security Fund (NSSF) for family coverage and sickness coverage. However, they are entitled to social security benefits only if:

- They have valid work permits, and
- Their countries of origin offer equal treatment to Lebanese workers (i.e., there are reciprocal social security arrangements).

Currently, only France, Italy, Belgium, and Britain have such reciprocal arrangements with Lebanon.

Foreign workers who do not meet these conditions, including Syrians, must still be registered and contribute to the NSSF but do not receive the benefits of such coverage.

Therefore, foreign workers are entitled to social security benefits only if their country of origin has reciprocal social security agreements with Lebanon and they hold valid work permits.

Summary:
- Foreign workers with valid work permits and from countries with reciprocal agreements (France, Italy, Belgium, Britain) are entitled to social security benefits.
- Foreign workers without such reciprocal agreements (e.g., Syrians) are not entitled to benefits despite contributing.
- Foreigners without valid work permits or not working for a specific employer are excluded from coverage.

This is in accordance with Article 9(3) and (5) of the NSSF law and related provisions.

**Sources:** data\Employment-Rights-Guide.pdf (Page 14), data\Self-Employment-Guide.pdf (Page 27), data\Employment-Rights-Guide.pdf (Page 34)

---

### Question
Are employees protected from dismissal while on leave?

**Answer:**

The guide does not provide this information.

**Sources:** 

---

### Question
What are the procedures for challenging abusive dismissal?

**Answer:**

Procedures for Challenging Abusive Dismissal:

1. Filing a Claim:
   - Workers who believe they have been terminated abusively must file a claim before the Labour Arbitration Council in their governorate.
   - The claim must be filed within 30 days (one month) from the date of termination.

2. Labour Arbitration Council:
   - The Council handles cases related to dismissal, labour accidents, wages, and other disputes under the law.
   - There is no cost for filing claims before the Council.

3. Decision Timeline:
   - The Labour Arbitration Council is required to reach a decision within three (3) months from the filing of the claim.

4. Outcome if Dismissal is Unjustified:
   - If the Council finds that the dismissal was not justified, the employer must reinstate the employee.

References:
- Article 50(B) and 50(E) of the Labour Code
- Labour Arbitration Councils' role and procedures as described in the guide

**Sources:** data\Employment-Rights-Guide.pdf (Page 19), data\Employment-Rights-Guide.pdf (Page 24), data\Employment-Rights-Guide.pdf (Page 20)

---

### Question
Does the guide regulate cryptocurrency salary payments?

**Answer:**

The guide does not provide this information.

**Sources:** 

---

### Question
Does the guide contain rules about remote work or work-from-home policies?

**Answer:**

The guide does not provide this information.

**Sources:** 

---

### Question
How does the guide define workers not linked to an employer?

**Answer:**

The guide defines workers not linked to an employer as those who are not connected in any form to an employer. Their coverage under the Social Security Law is determined by a decree issued by the Council of Ministers upon a specific procedure. Currently, workers who are not linked to any employer are excluded from the Social Security Law coverage. 

This category includes workers who may work independently or on an ad hoc basis, such as daily workers who are not in an employment relationship as defined by the Labour Law but work as contractors on short-term service contracts.

**Sources:** data\Self-Employment-Guide.pdf (Page 27), data\Employment-Rights-Guide.pdf (Page 10), data\Self-Employment-Guide.pdf (Page 26)

---

### Question
Are self-employed individuals covered by social security?

**Answer:**

Answer:

- Self-employed individuals, both Lebanese and foreigners, are currently excluded from the formal social security system.
- Although Lebanese self-employed workers had some access to social security in 2003, this is no longer the case.
- The Social Security Law of 1963, which established the National Social Security Fund (NSSF), was initially intended to cover all formal sector workers, but its application remains incomplete and fragmented.
- Workers not linked to any employer, including self-employed persons, are still excluded from social security coverage at present.
- Foreigners working in Lebanon can access social security benefits only if they have valid work permits and their countries of origin provide equal treatment to Lebanese workers; however, those not working for a specific employer are excluded.

Therefore, self-employed individuals are generally not covered by social security under the current legal framework.

**Sources:** data\Self-Employment-Guide.pdf (Page 24), data\Self-Employment-Guide.pdf (Page 27), data\Self-Employment-Guide.pdf (Page 26)

---

### Question
Are foreigners eligible for social security benefits?

**Answer:**

Answer:

Yes, foreigners working in Lebanon are entitled to social security benefits provided they meet the following conditions:

1. They have valid work permits.
2. Their countries of origin offer equal treatment to Lebanese workers.

However, foreigners who do not work for a specific employer are excluded from coverage under the Social Security Law. Additionally, while foreigners must pay contributions for family and sickness coverage, they do not benefit from these allowances. Syrian employees do not benefit from the end-of-service indemnity.

This information is based on the Social Security Law and related provisions as outlined in the guide.

**Sources:** data\Employment-Rights-Guide.pdf (Page 14), data\Self-Employment-Guide.pdf (Page 27), data\Self-Employment-Guide.pdf (Page 26)

---

### Question
When is a daily worker considered an employee versus self-employed?

**Answer:**

A daily worker is considered an employee if they work under the subordination and supervision of an employer for a wage, even for a short time. Conversely, a daily worker is considered self-employed if they are not in an ‘employment relationship’ as defined in the Labour Law, for example, if they work on an ad hoc basis as contractors on a short-term service contract. Additionally, daily workers who work on a small scale for themselves and undertake small commerce or simple trades with low general expenses are not considered traders. 

This distinction is based on the presence or absence of an employment relationship and the nature of the work arrangement.

**Sources:** data\Self-Employment-Guide.pdf (Page 24), data\Self-Employment-Guide.pdf (Page 27), data\Self-Employment-Guide.pdf (Page 21)

---

### Question
Must individuals conducting commercial activities register as traders?

**Answer:**

Based on the provided context:

- Individuals conducting commercial activities are called traders and must abide by the general regulations of the Commercial Code (CoC).
- Physical persons can register themselves as traders at the Commercial Register of Beirut or the locality where their main business is located.
- The registration procedure involves submitting an application and declaration form, providing one photo, paying fees (approximately LBP 2,500,000), and finalizing the procedure at the Commercial Register.
- However, persons who are not formally registered as traders in the Commercial Register are still considered to be traders under the law.

Therefore:

**Individuals conducting commercial activities are not strictly required to register as traders to be considered traders under the law. Registration is available and involves a formal procedure, but lack of registration does not exempt a person from being legally recognized as a trader.**

**Sources:** data\Self-Employment-Guide.pdf (Page 30), data\Self-Employment-Guide.pdf (Page 29)

---

### Question
What activities are considered commercial in nature?

**Answer:**

Based on the provided context, the activities considered commercial in nature in Lebanon include:

1. Purchasing goods and other tangible and intangible movables for the purpose of selling them for a profit, whether sold as they were or after being converted or transformed.
2. Renting such goods.
3. Manufacturing and factories, even if associated with agricultural investment, unless the transformation is done by simple material work.
4. Mining and petroleum activities.
5. Public warehouses.
6. Supply of products.
7. Printing and publishing.
8. Exchange and banking.
9. Land and sea transport.
10. Public entertainment.
11. All activities equivalent to the above due to similarity in characteristics and purposes.

These activities are outlined in Articles 6 and 7 of the Commercial Register (CoC).

**Sources:** data\Self-Employment-Guide.pdf (Page 28), data\Self-Employment-Guide.pdf (Page 7), data\Self-Employment-Guide.pdf (Page 31)

---

### Question
Are workers who work for multiple employers covered by social security?

**Answer:**

Based on the provided context:

- The Social Security Law mentions workers who work for several employers, explicitly naming workers at sea, in ports, construction workers, and workers not linked in any form to an employer, provided their coverage is determined by a decree issued by the Council of Ministers upon a specific procedure.  
- However, the application of the Social Security Law remains incomplete and fragmented.  
- Workers who are not linked to any employer are still excluded from the law at present.

Therefore, workers who work for multiple employers may be covered by social security only if a specific decree by the Council of Ministers determines their coverage. Otherwise, coverage is not guaranteed.

Summary:  
Workers working for multiple employers are potentially covered by social security if a decree by the Council of Ministers establishes such coverage; otherwise, they are currently excluded.

**Sources:** data\Self-Employment-Guide.pdf (Page 27), data\Self-Employment-Guide.pdf (Page 26)

---

### Question
Are self-employed individuals required to pay tax?

**Answer:**

Answer:

Yes, self-employed individuals in Lebanon are required to pay tax. According to the Income Tax Act and related laws, any person working as a self-employed person (whether Lebanese or foreigner) must register with the Tax Administration and abide by Lebanese tax laws. The tax is imposed on the total income generated through any work or business activity exercised in Lebanon. Failure to declare income may result in legal fines for undeclared income.

**Sources:** data\Self-Employment-Guide.pdf (Page 24), data\Self-Employment-Guide.pdf (Page 32)

---

### Question
Are self-employed individuals required to pay municipal fees, and on what basis are they calculated?

**Answer:**

Answer:

Yes, self-employed individuals are required to pay municipal fees. These fees depend on the type of business and are paid to the municipality where the business is established.

The basis for calculating these fees includes:

1. Rental Value: An annual fee charged on the rental value of the place occupied by the business, whether the individual is a tenant or owner. This applies broadly to any fixed place used for housing, trade, industry, or other purposes, including kiosks and vehicles permanently fixed on private land, as well as vacant lands used for non-agricultural investment.

2. Meeting Venues and Gambling Clubs: License fees (collected once upon granting the license) and annual investment fees apply to establishments such as restaurants, fast food premises, movie theatres, coffee shops, hotels, game centres, and similar touristic premises.

3. Advertisements: Fees (license and/or investment fees) apply to advertisements related to businesses, goods, and services within the municipal area, regardless of type or permanence.

4. Usage of Municipal Public Land: License and investment fees are due for using municipal public land for business purposes, such as establishing kiosks, stands, or advertising signs.

5. Classified Businesses: Licensing fees collected once upon granting the license apply to industrial institutions or shops classified by law based on their type and environmental or neighborhood risk.

It is important to note that these municipal fees are additional to any initial licenses required under special laws.

**Sources:** data\Self-Employment-Guide.pdf (Page 34)

---

### Question
Are small-scale daily labourers considered traders under commercial law?

**Answer:**

Based on the provided context:

- Traders are defined as persons whose career is to undertake commercial activities regularly, independently, and as a main source of livelihood.
- Small-scale daily labourers, such as ambulant merchants or those involved in small-scale commerce with low general expenses, are exempted under Article 10 of the Commercial Code (CoC).
- These small-scale traders are considered traders in terms of their activities being commercial, but due to the exemption, they are not subject to obligations like keeping commercial books or registering at the Commercial Register.
- The exemption applies to persons undertaking small commerce or simple trades on a daily basis, including ambulant merchants and small transportation service owners.
- However, the context does not explicitly mention "small-scale daily labourers" outside of commerce or trade activities.

Conclusion:
Small-scale daily labourers engaged in small-scale commerce or simple trades are considered traders under commercial law but benefit from an exemption under Article 10 of the CoC. If the daily labourers are not engaged in commercial activities as defined, the guide does not provide information on their status.

Therefore, **small-scale daily labourers involved in small-scale commerce or simple trades are considered traders but are exempt from certain commercial obligations.**

**Sources:** data\Self-Employment-Guide.pdf (Page 30), data\Self-Employment-Guide.pdf (Page 29)

---

### Question
Does the guide set minimum pricing standards for freelance services?

**Answer:**

The guide does not provide this information.

**Sources:** 

---

### Question
Does the guide provide government grants or startup funding schemes for self-employed workers?

**Answer:**

The guide does not provide this information.

**Sources:** 

---

## Evaluation

In [61]:
ground_truth = {

    # EMPLOYMENT RIGHTS GUIDE 

    "What is the legally mandated number of annual leave days?": {
        "document": "Employment-Rights-Guide.pdf",
        "pages": [16, 24]
    },

    "How does annual leave increase based on years of service?": {
        "document": "Employment-Rights-Guide.pdf",
        "pages": [16]
    },

    "What is the official minimum wage?": {
        "document": "Employment-Rights-Guide.pdf",
        "pages": [13]
    },

    "What are the maximum legal working hours per week?": {
        "document": "Employment-Rights-Guide.pdf",
        "pages": [15]
    },

    "How is overtime compensated?": {
        "document": "Employment-Rights-Guide.pdf",
        "pages": [15]
    },

    "What are the rules governing sick leave?": {
        "document": "Employment-Rights-Guide.pdf",
        "pages": [17]
    },

    "Is maternity leave paid and for how long?": {
        "document": "Employment-Rights-Guide.pdf",
        "pages": [17]
    },

    "What rights exist during the probation period?": {
        "document": "Employment-Rights-Guide.pdf",
        "pages": [12]
    },

    "What prior notice is required before termination?": {
        "document": "Employment-Rights-Guide.pdf",
        "pages": [21]
    },

    "What compensation is due in case of abusive dismissal?": {
        "document": "Employment-Rights-Guide.pdf",
        "pages": [21]
    },

    "Under what conditions can employment be terminated due to force majeure?": {
        "document": "Employment-Rights-Guide.pdf",
        "pages": [22, 23]
    },

    "What are the employer's obligations regarding social security contributions?": {
        "document": "Employment-Rights-Guide.pdf",
        "pages": [14, 15, 34]
    },

    "Are foreign workers entitled to social security benefits?": {
        "document": "Employment-Rights-Guide.pdf",
        "pages": [15, 34]
    },

    "Are employees protected from dismissal while on leave?": {
        "document": "Employment-Rights-Guide.pdf",
        "pages": [16, 17, 24]
    },

    "What are the procedures for challenging abusive dismissal?": {
        "document": "Employment-Rights-Guide.pdf",
        "pages": [24]
    },

    "Does the guide regulate cryptocurrency salary payments?": {
    "document": None,
    "pages": []
    },

    "Does the guide contain rules about remote work or work-from-home policies?": {
    "document": None,
    "pages": []
    },


    # SELF-EMPLOYMENT GUIDE
    "How does the guide define workers not linked to an employer?": {
        "document": "Self-Employment-Guide.pdf",
        "pages": [24]
    },

    "Are self-employed individuals covered by social security?": {
        "document": "Self-Employment-Guide.pdf",
        "pages": [26, 27]
    },

    "Are foreigners eligible for social security benefits?": {
        "document": "Self-Employment-Guide.pdf",
        "pages": [27]
    },

    "When is a daily worker considered an employee versus self-employed?": {
        "document": "Self-Employment-Guide.pdf",
        "pages": [27]
    },

    "Must individuals conducting commercial activities register as traders?": {
        "document": "Self-Employment-Guide.pdf",
        "pages": [29, 30]
    },

    "What activities are considered commercial in nature?": {
        "document": "Self-Employment-Guide.pdf",
        "pages": [28, 29]
    },

    "Are workers who work for multiple employers covered by social security?": {
        "document": "Self-Employment-Guide.pdf",
        "pages": [26, 27]
    },

    "Are self-employed individuals required to pay tax?": {
        "document": "Self-Employment-Guide.pdf",
        "pages": [32, 33]
    },

    "Are self-employed individuals required to pay municipal fees, and on what basis are they calculated?": {
        "document": "Self-Employment-Guide.pdf",
        "pages": [34]
    },

    "Are small-scale daily labourers considered traders under commercial law?": {
        "document": "Self-Employment-Guide.pdf",
        "pages": [29, 30]
    },

    "Does the guide set minimum pricing standards for freelance services?": {
    "document": None,
    "pages": []
    },

    "Does the guide provide government grants or startup funding schemes for self-employed workers?": {
    "document": None,
    "pages": []
    },
}

In [62]:
# check for missing questions
for q in evaluation_questions:
    if q not in ground_truth:
        print("Missing in ground_truth:", q)

#### Precision@K

In [63]:
def compute_precision_at_k(evaluation_questions, ground_truth, k=3):
    total_retrieved = 0
    total_relevant = 0

    for question in evaluation_questions:

        gt = ground_truth[question]
        gt_doc = gt["document"]
        gt_pages = set(str(p) for p in gt["pages"])

        retrieved_docs = retrieve(question, k=k)

        # Remove duplicate (doc, page) pairs
        retrieved_pairs = set(
            (
                doc["source"].split("\\")[-1],
                str(doc["page"])
            )
            for doc in retrieved_docs
        )

        total_retrieved += len(retrieved_pairs)

        for doc_name, page in retrieved_pairs:
            if gt_doc is not None and doc_name == gt_doc and page in gt_pages:
                total_relevant += 1

    return total_relevant / total_retrieved if total_retrieved > 0 else 0


precision_at_3 = compute_precision_at_k(
    evaluation_questions,
    ground_truth,
    k=3
)

print(f"Precision@3: {precision_at_3:.2f}")


Precision@3: 0.39


#### Recall@K

In [65]:
def compute_recall_at_k(evaluation_questions, ground_truth, k=3):
    total_questions = 0
    questions_with_hit = 0

    for question in evaluation_questions:

        gt = ground_truth[question]
        gt_doc = gt["document"]
        gt_pages = set(str(p) for p in gt["pages"])

        if gt_doc is None:
            continue  # skip negative questions for recall

        total_questions += 1

        retrieved_docs = retrieve(question, k=k)

        retrieved_pairs = set(
            (
                doc["source"].split("\\")[-1],
                str(doc["page"])
            )
            for doc in retrieved_docs
        )

        for doc_name, page in retrieved_pairs:
            if doc_name == gt_doc and page in gt_pages:
                questions_with_hit += 1
                break

    return questions_with_hit / total_questions if total_questions > 0 else 0

recall_at_3 = compute_recall_at_k(
    evaluation_questions,
    ground_truth,
    k=3
)

print(f"Recall@3: {recall_at_3:.2f}")

Recall@3: 0.96


#### Hit@K

In [66]:
def compute_hit_at_k(evaluation_questions, ground_truth, k=3):
    hits = 0
    total = 0

    for question in evaluation_questions:

        gt = ground_truth[question]
        gt_doc = gt["document"]
        gt_pages = set(str(p) for p in gt["pages"])

        if gt_doc is None:
            continue

        total += 1

        retrieved_docs = retrieve(question, k=k)

        for doc in retrieved_docs:
            doc_name = doc["source"].split("\\")[-1]
            page = str(doc["page"])

            if doc_name == gt_doc and page in gt_pages:
                hits += 1
                break

    return hits / total if total > 0 else 0

hit_at_3 = compute_hit_at_k(
    evaluation_questions,
    ground_truth,
    k=3
)

print(f"Hit@3: {hit_at_3:.2f}")

Hit@3: 0.96


In [67]:
for k in [1, 3, 5]:
    print(f"\nK = {k}")
    print(f"Hit@{k}: {compute_hit_at_k(evaluation_questions, ground_truth, k=k):.2f}")
    print(f"Recall@{k}: {compute_recall_at_k(evaluation_questions, ground_truth, k=k):.2f}")


K = 1
Hit@1: 0.76
Recall@1: 0.76

K = 3
Hit@3: 0.96
Recall@3: 0.96

K = 5
Hit@5: 0.96
Recall@5: 0.96


### Retrieval Evaluation Summary:

**Metrics Used:**

- **Precision@K** - Proportion of retrieved pages in the top K that are relevant. Measures retrieval cleanliness (noise level).

- **Recall@K** - Measures whether the retriever successfully retrieves the relevant ground-truth page(s) within the top K results.

- **Hit@K** - Measures whether at least one correct source page is retrieved in the top K results.

**Interpretation:**

- At K=1, the retriever returns the correct page as the top result in 76% of cases.

- At K=3, relevant pages are retrieved in 96% of questions.

- Increasing K to 5 does not improve performance, indicating that relevant pages are already retrieved within the top 3.

**Conclusion:**

The retriever performs strongly.

While Precision@3 was lower due to extra non-relevant pages being retrieved, Hit@K and Recall@K demonstrate that the system reliably retrieves the correct source context for nearly all questions.

### LLM Answer Evaluation:

- Total Evaluation Questions: 29

- Retrieval Top-k: 3

- LLM Model: gpt-4.1-mini

**Observations:**

- The system produces structured and grounded legal answers.

- Irrelevant questions are correctly rejected with:

    "The guide does not provide this information."

- No hallucinated legal numbers were observed during evaluation.

This indicates effective retrieval grounding and prompt control.

### Limitations

- Evaluation conducted on a relatively small, manually selected dataset.

- Retrieval relies solely on dense vector search (no reranking).

- Fixed top-k retrieval strategy without adaptive selection..

- Evaluation focuses on retrieval metrics; answer quality assessed manually.

### Future Improvements

- Introduce reranking for improved precision.

- Implement hybrid search (BM25 + dense retrieval).

- Experiment with adaptive top-K selection.

- Add automated answer-level evaluation (e.g., LLM-as-judge or semantic similarity scoring).